Drop support for running etcd in pod network

giorio94 opened the pull request:
Support for running etcd in pod network has been officially deprecated in v1.16 \[1], and disabled by default in v1.17 \[2]. It is now time to drop it completely, to allow simplifying the watchers logic and possible future associated refactoring.

\[1]: https://github.com/cilium/cilium/commit/f99f10bd8843d70e208b03eae9c556f64a143204 ("docs: Deprecate support for podnetwork etcd")
\[2]: https://github.com/cilium/cilium/commit/86c85da1cc3d5a945c1dbac7d6e9ec3bde9f5a63 ("watchers: disable CRD to kvstore handover logic")

Please review commit by commit, and refer to the individual commit messages for additional details.

I'm not adding this removal to the upgrade notes, as already announced as part of the previous release, when this support has been disabled behind a hidden flag.

<!-- Description of change -->

```release-note
Remove deprecated and disabled by default support for running the Cilium KVStore in pod network
```

giorio94 committed:
watchers: drop CRD to kvstore handover logic from CN watcher

Following the removal of the support for running etcd in pod network,
let's rework the CiliumNode watcher dropping the handover logic, which
is complex and fragile. Indeed, the watcher is now guaranteed to not
be started at all when Cilium is configured in kvstore mode. For the
same reason, it is no longer necessary to mark the CN resource as
stoppable. While being there, let's align the structure to that of
the other watchers.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
watchers: drop CRD to kvstore handover logic from CEP/CES watchers

Following the removal of the support for running etcd in pod network,
let's rework the CiliumEndpoint and CiliumEndpointSlice watchers
dropping the handover logic, which is complex and fragile. Indeed,
the watchers are now guaranteed to not be started at all when Cilium
is configured in kvstore mode. For the same reason, it is no longer
necessary to mark the CEP/CES resources as stoppable. While being
there, let's align the structure to that of the other watchers.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
watchers: drop CRD to kvstore handover logic from pod watchers

Following the removal of the support for running etcd in pod network,
let's rework the Pod watchers dropping the handover logic (that is
switching from watching all pods to local pods only), in favor of
always watching the local pods only. While being there, let's also
align the structure to that of the other watchers.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: drop unused Connected and Disconnected methods

They are no longer used following the removal of the CRD to kvstore
handover logic.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: drop support for running etcd in pod network

Support for running etcd in pod network has been officially deprecated
in v1.16 [1], and disabled by default in v1.17 [2]. It is now time to
drop it completely, to allow simplifying the watchers logic and possible
future associated refactoring.

[1]: f99f10bd8843 ("docs: Deprecate support for podnetwork etcd")
[2]: 86c85da1cc3d ("watchers: disable CRD to kvstore handover logic")

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 commented on pkg/k8s/watchers/pod.go:
```diff
@@ -164,164 +150,58 @@ func hostPortServiceName(pod *slim_corev1.Pod, port uint16, cluster string) *loa
-	if !option.Config.DisableCiliumEndpointCRD || option.Config.KVstoreEnabled() {
-		watchNodePods()
-		return
-	}
```
There seem to be currently two possibilities in which the CiliumEndpoint CRD may be disabled by the user, hence falling back here to watching all pods rather than the local ones only:

1. Cilium is running in kvstore mode; in this case, the fallback logic is not necessary, and we can drop it (it is also already currently disabled), because we rely on the information retrieved from etcd.
2. NetworkPolicy support is disabled (related: 19fe642a4fa8ed8c8a5980c97cbd633f7a1ea2c1). @dlapcevic Could you please confirm whether it is necessary to watch all pods in this case? If so, I'll have to rework this commit to preserve the current logic for that use-case. FWIW, I'm not seeing any CI failure that may indicate problems, but I'm unsure about the coverage in that area.

marseel commented on pkg/k8s/resource_ctors.go:
```diff
@@ -415,7 +414,6 @@ func CiliumSlimEndpointResource(params CiliumResourceParams, _ *node.LocalNodeSt
 		}, TransformToCiliumEndpoint),
 		resource.WithMetric("CiliumEndpoint"),
 		resource.WithIndexers(indexers),
-		resource.WithStoppableInformer(),
```
should we delete `WithStoppableInformer`? I don't think it's used anymore.

giorio94 replied:
Dropped via the last commit, that reverts the one that initially introduced that support.

marseel commented on daemon/cmd/daemon_main.go:
```diff
@@ -503,10 +503,6 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Duration(option.KVstoreConnectivityTimeout, defaults.KVstoreConnectivityTimeout, "Time after which an incomplete kvstore operation  is considered failed")
 	option.BindEnv(vp, option.KVstoreConnectivityTimeout)
 
-	flags.Bool(option.KVstorePodNetworkSupport, defaults.KVstorePodNetworkSupport, "Enable the support for running the Cilium KVstore in pod network")
```
Let's maybe add it to upgrade notes one more time to ensure that if someone was configuring this flag in v1.17 is aware that now it's noop?

tommyp1ckles replied:
does passing a non existent flag act as a noop, or would that produce an error on its own?

giorio94 replied:
> Let's maybe add it to upgrade notes one more time to ensure that if someone was configuring this flag in v1.17 is aware that now it's noop?

I had thought about adding a release note, but ended up not doing that to avoid possible confusion. Indeed, the support removal has already been announced in v1.17, and this flag has always been hidden and undocumented. Happy to add a note if you feel that would be more appropriate though. 

> does passing a non existent flag act as a noop, or would that produce an error on its own?

It depends. It triggers a hard error if passed as a command line argument due to the flag not existing anymore, while it is silently ignored if specified via an environment variable or config map entry.

tommyp1ckles commented on pkg/k8s/watchers/cilium_node.go:
```diff
@@ -65,104 +63,54 @@ type K8sCiliumNodeWatcher struct {
+
+	k.k8sResourceSynced.BlockWaitGroupToSyncResources(
+		ctx.Done(),
+		nil,
```
why do we need this if we don't pass a wait group?

giorio94 replied:
Yes, the mention to the wait group is a bit misleading, but that function registers the synced func and eventually marks the given resource as synchronized. That is used in the daemon initialization to block until all appropriate informers are synchronized.

https://github.com/cilium/cilium/blob/1914c208c0a9dedb33e6e046265eabd87d5e532c/daemon/cmd/daemon_main.go#L1675-L1683

giorio94 committed:
Revert "resource: Add support for releasable resources"

This reverts commit 42267013439651a02ac470581dd7821103fd077f.

Now that all users of releasable resources have been removed, let's drop
the extra logic to stop the underlying informer as well, to keep things
as simple as possible.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
bgp: drop no-op store release

Currently, the `cleanupStatus` method eventually releases the BGP peer
config store. However, that resource is not initialized as stoppable,
hence this operation is a no-op. Nevertheless, the informer could not
be stopped, as events are still observed. In preparation for the
subsequent removal of the support for releasable resources, let's
drop this call.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

pippolo84 reviewed (approved):
Nice! :heavy_check_mark: 

Since we removed the releasable resources, we could drop `promise.MapError` as well, but let's do that in a separate PR. :+1:

giorio94 committed:
Revert "promise: Add MapError operator"

This reverts commit 49ae544f210c6845ca731b18e8a0d9c32abd73d2.

Now that all users of `promise.MapError` have been removed, let's drop
it as well.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>
